src/BSON/ArrayList.c

```diff
@@ -0,0 +1,390 @@
+{
+	php_phongo_arraylist_t* intern;
+	HashTable*              props;
+
+	intern = Z_OBJ_ARRAYLIST(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 1);
+
+	if (!intern->bson) {
+		return props;
+	}
+
+	{
+		zval data;
+
+		ZVAL_STR(&data, php_base64_encode((const unsigned char*) bson_get_data(intern->bson), intern->bson->len));
+		zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+	}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_ArrayList)
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, fromPHP)
+{
+	zval                    zv;
+	php_phongo_arraylist_t* intern;
+	zval*                   data;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_ARRAY(data)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	if (!zend_array_is_list(Z_ARRVAL_P(data))) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected value to be a list, but given array is not.");
+		return;
+	}
+
+	object_init_ex(&zv, php_phongo_arraylist_ce);
+	intern = Z_ARRAYLIST_OBJ_P(&zv);
+
+	intern->bson = bson_new();
+	php_phongo_zval_to_bson(data, PHONGO_BSON_NONE, intern->bson, NULL);
+
+	RETURN_ZVAL(&zv, 1, 1);
+}
+
+static bool seek_iter_to_index(bson_iter_t* iter, zend_long index)
+{
+	for (zend_long i = 0; i <= index; i++) {
+		if (!bson_iter_next(iter)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, get)
+{
+	php_phongo_arraylist_t* intern;
+	zend_long               index;
+	bson_iter_t             iter;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_LONG(index)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	intern = Z_ARRAYLIST_OBJ_P(getThis());
+	if (!bson_iter_init(&iter, intern->bson)) {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME, "Could not initialize BSON iterator.");
+	}
+
+	if (!seek_iter_to_index(&iter, index)) {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME, "Could not find index \"%d\" in BSON data", index);
+		RETURN_NULL();
+	}
+
+	php_phongo_bson_iter_to_zval(return_value, &iter);
+}
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, getIterator)
+{
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	phongo_iterator_init(return_value, getThis());
+}
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, has)
+{
+	php_phongo_arraylist_t* intern;
+	zend_long               index;
+	bson_iter_t             iter;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_LONG(index)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	intern = Z_ARRAYLIST_OBJ_P(getThis());
+	if (!bson_iter_init(&iter, intern->bson)) {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME, "Could not initialize BSON iterator.");
+	}
+
+	RETURN_BOOL(seek_iter_to_index(&iter, index));
+}
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, toPHP)
+{
+	php_phongo_arraylist_t* intern;
+	zval*                   typemap = NULL;
+	php_phongo_bson_state   state;
+
+	PHONGO_PARSE_PARAMETERS_START(0, 1)
+	Z_PARAM_OPTIONAL
+	Z_PARAM_ARRAY(typemap)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	PHONGO_BSON_INIT_STATE(state);
```

Comment:
Re: https://github.com/mongodb/mongo-php-driver/pull/1387#discussion_r1045524873

Where does the root element default to an array?

Reply:
The change was made in php_phongo.c where we use `PHONGO_TYPEMAP_CLASS` for the root type if it was `PHONGO_TYPEMAP_NONE`. That logic now checks the state to see if it is visiting an array and uses `PHONGO_TYPEMAP_NATIVE_ARRAY ` if so: https://github.com/mongodb/mongo-php-driver/pull/1387/commits/7209833eb33d4019c21d459aa99265ef307ba625#diff-09bc6eb8abd9dcb66a40cd127006f752677a35440667f47d505cd9474abaff43R1074